hooks/pre-commit

```diff
@@ -1,29 +1,125 @@
+
+set -euo pipefail
+
+source "tools/codestyle/globals.source"
+
+function path_under_any_of_dirs()
+{
+    local path="$1" dir
+    shift
+    for dir in "$@"; do
+        if [[ "${path}" == "${dir}/"* ]]; then
+            return 0
+        fi
+    done
+    return 1
+}
+
+# Files with staged changes that are required to follow the coding style.
+function get_staged_files()
+{
+    local tidy_globs=() dir ext
+    for dir in "${TIDY_DIRS[@]}"; do
+        for ext in "${TIDY_EXTENSIONS[@]}"; do
+            tidy_globs+=("${dir}/*.${ext}")
+        done
+    done
+
+    local untidy_dirs=() untidy_path
+    while IFS= read -r -d '' untidy_path; do
+        untidy_dirs+=("${untidy_path%/${UNTIDY_FILE}}")
+    done < <(git ls-files -z -- "*/${UNTIDY_FILE}")
+
+    STAGED_FILES=()
+
+    local file
+    while IFS= read -r -d '' file; do
+        if ! path_under_any_of_dirs "${file}" "${untidy_dirs[@]}"; then
+            STAGED_FILES+=("${file}")
+        fi
+    done < <(git diff -z --name-only --cached -- "${tidy_globs[@]}")
+
+    NUM_STAGED="${#STAGED_FILES[@]}"
+}
+
+# Files from STAGED_FILES that also have unstaged changes. Beware, if
+# STAGED_FILES is empty then this will return all unstaged files.
+function get_unstaged_files()
+{
+    UNSTAGED_FILES=()
+
+    while IFS= read -r -d '' unstaged_file; do
+        UNSTAGED_FILES+=("${unstaged_file}")
+    done < <(git diff -z --name-only -- "${STAGED_FILES[@]}")
+
+    NUM_UNSTAGED="${#UNSTAGED_FILES[@]}"
+}
+
+exit_status=0
+
+get_staged_files
+
+if ((${NUM_STAGED} == 0)); then
+    echo >&2 "$0: Nothing to do."
+    exit 0
+fi
+
+get_unstaged_files
+
+echo >&2 "$0: Tidying staged files..."
+
+# Avoid overwriting unstaged changes.
+for file in "${UNSTAGED_FILES[@]}"; do
+    mv "${file}" "${file}.unstaged"
+    git checkout -- "${file}"
+done
+
+# Tidy the staged version of each file.
+STAGED_PIDS=()
+for file in "${STAGED_FILES[@]}"; do
+    "tools/codestyle/tidy_file.sh" "${file}" & # Run in background. Must not use Git.
+    STAGED_PIDS+=( $! ) # store process ID
+done
+
+# Wait for background processes to finish and re-stage the tidy files.
+for idx in "${!STAGED_PIDS[@]}"; do
+    if wait "${STAGED_PIDS[${idx}]}"; then
+        # Process was successful
+        file="${STAGED_FILES[${idx}]}"
+        git add -- "${file}" # Run in foreground to avoid git index conflicts.
+    else
+        # Process failed
+        status=$?
+        exit_status=${status}
+    fi
+done
+
+if ((${NUM_UNSTAGED} == 0)); then
+    echo >&2 "$0: Tidying complete!"
+    exit ${exit_status}
+fi
+
+echo >&2 "$0: Restoring and tidying unstaged changes..."
+
+# Restore and tidy the unstaged version to reduce diff with staged version.
+UNSTAGED_PIDS=()
+for file in "${UNSTAGED_FILES[@]}"; do
+    mv "${file}.unstaged" "${file}"
+    "tools/codestyle/tidy_file.sh" "${file}" & # Run in background. Must not use Git.
+    UNSTAGED_PIDS+=( $! ) # store process ID
+done
```

Comment:
If we want extreme parallelism, isn't it theoretically possible to do this in parallel with the staged files? I think you could do that basically by reordering the blocks a bit, like this:

```bash
# Tidy the staged version of each file.
STAGED_PIDS=()
for file in "${STAGED_FILES[@]}"; do
    "tools/codestyle/tidy_file.sh" "${file}" & # Run in background. Must not use Git.
    STAGED_PIDS+=( $! ) # store process ID
done

# Tidy the unstaged version to reduce diff with staged version.
UNSTAGED_PIDS=()
for file in "${UNSTAGED_FILES[@]}"; do
    # TODO: make tidy_file.sh understand the `.unstaged` extension
    "tools/codestyle/tidy_file.sh" "${file}.unstaged" & # Run in background. Must not use Git.
    UNSTAGED_PIDS+=( $! ) # store process ID
done

# Wait for background processes to finish and re-stage the tidy files.
for idx in "${!STAGED_PIDS[@]}"; do
    if wait "${STAGED_PIDS[${idx}]}"; then
        # Process was successful
        file="${STAGED_FILES[${idx}]}"
        git add -- "${file}" # Run in foreground to avoid git index conflicts.
    else
        # Process failed
        status=$?
        exit_status=${status}
    fi
done

# Wait for each background process to finish and restore the unstaged files
for idx in "${!UNSTAGED_PIDS[@]}"; do
    if wait "${UNSTAGED_PIDS[${idx}]}"; then
        mv "${file}.unstaged" "${file}"
    else
        # Process failed
        exit_status=$?
    fi
done

Reply:
Good idea!